From 9a9c776d4b440081efbc16e0669cc9ec19b955db Mon Sep 17 00:00:00 2001
From: zhengq <qi.zheng@intel.com>
Date: Tue, 28 Apr 2015 00:54:42 +0800
Subject: [PATCH] Pinctrl: set TX/RX both enabled for VGPIO6

Special treatment for virtual GPIO 6.
This GPIO is treated as input interrupt. Generally, only the
RX will be enabled for inpt GPIO. But to make it capable of
generating interrupt by PMC FW, the TX should be enabled as
well.

Tracked-On: https://jira01.devtools.intel.com/browse/AKP-1500
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-15649
Change-Id: Ic1250e37daf92a761f9557ecfb63570f7593f0c7
Signed-off-by: zhengq <qi.zheng@intel.com>
---
 drivers/pinctrl/pinctrl-cherryview.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/pinctrl/pinctrl-cherryview.c b/drivers/pinctrl/pinctrl-cherryview.c
index 4f1f214d8829..ca30f3bd1f68 100644
--- a/drivers/pinctrl/pinctrl-cherryview.c
+++ b/drivers/pinctrl/pinctrl-cherryview.c
@@ -951,9 +951,16 @@ static int chv_gpio_direction_input(struct gpio_chip *chip, unsigned offset)
 
 	spin_lock_irqsave(&cg->lock, flags);
 
+	/*Special workaround for virtual GPIO 6. This GPIO need be
+	*configured as tx/rx both enabled even for input to support
+	*interrupt triggered on PMC firmware. For general input gpio,
+	*enable rx only*/
 	value = chv_readl(reg) & (~CV_GPIO_CFG_MASK);
-	/* Disable TX and Enable RX */
-	value |= CV_GPIO_RX_EN;
+	if (cg->chip.ngpio != 8 || offset != 6) {
+		/* Disable TX and Enable RX */
+		value |= CV_GPIO_RX_EN;
+	}
+
 	chv_writel(value, reg);
 
 	spin_unlock_irqrestore(&cg->lock, flags);
-- 
1.9.1

